### ISSUE TITLE
Ansible remote provisioning always connect to localhost

### ISSUE THREAD

**Issue by vankhoa011** [mitchellh/packer#3564]
I used Packer to create AMI ( amazon-ebs  ) and Ansible remote for provisioning.
However, Packer always connect to localhost .

The log file: 
https://gist.github.com/vankhoa011/a6800c130b8fecdb3de12bcf48a50036

Packer version: 0.10.1
Packer template:
```
{
  "builders": [{
    "type": "amazon-ebs",
    "region": "us-east-1",
    "source_ami": "ami-a1fa1acc",
    "instance_type": "t2.micro",
    "ssh_username": "ec2-user",
    "ami_name": "ecs {{timestamp}}",
    "vpc_id": "vpc-0a38056f",
    "subnet_id": "subnet-b829a293",
    "associate_public_ip_address": "true"
  }],
  "provisioners": [{
    "type": "ansible",
    "playbook_file": "./playbook.yml",
    "extra_arguments": "-vvvv",
    "user": "ec2-user",
    "host_alias": "default",
    "ansible_env_vars": ["ANSIBLE_HOST_KEY_CHECKING=False", "ANSIBLE_SSH_ARGS='-o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=60s'", "ANSIBLE_NOCOLOR=True"]
  }]
}
```
playbook.yml

```
- hosts: default
  gather_facts: True
  user: ec2-user
  sudo: True
  roles:
    - logstash-forwarder
    - logrotate
    - newrelic
```

**Comment by ricardclau** [comment]
Hi @vankhoa011 

We use
```
hosts: all 
````

in playbook.yml and the ansible provisioner in Packer is just these 2 lines

```
 "provisioners": [
        {
            "type": "ansible",
            "playbook_file": "route_to_playbook.yml"
        }
]
```

Maybe you encountered some sort of bug combining extra parameters?

**Comment by rickard-von-essen** [comment]
This actually works for me (just dropped the vpc and subnet). 

Could you double check that your VPC don't have any ACL's that interfere with the communication from you control machine and the guest (the provisioned EC2).

The Packer Ansible (remote) provisioner is supposed to connect to localhost where a proxy forwards the traffic to the guest.

```
==> amazon-ebs: SSH proxy: serving on 127.0.0.1:51205
```

**Comment by vankhoa011** [comment]
@ricardclau : I changed a little bit as I mentioned above. However, still failed.  Because I don't know which ssh key, private ip, and user that the Ansible use to connect to remote host.  



@rickard-von-essen: Yes, of course . My control machine can connect to VPC. I create a instance which is in public subnet. 

I use SSH key which created by Packer, and connect successfully. 

Full log file .

[](https://gist.github.com/vankhoa011/39c64fb8bbb241238976a9e340558b35
)

**Comment by rickard-von-essen** [comment]
> Because I don't know which ssh key, private ip, and user that the Ansible use to connect to remote host.

The a one time ssh key is generated and is used by ansible to connect to the proxy. The proxy will connect with the credentials specified in your builder, in your case a temporary aws key pair with the `ec2-user`.

**Comment by vankhoa011** [comment]
@rickard-von-essen : But  the packer still can't connect to remote host . Do you have any clue ? 

```
2016/05/21 21:56:55 ui:     amazon-ebs: fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh", "unreachable": true}
    amazon-ebs: fatal: [default]: UNREACHABLE! => {"changed": false, "msg": "SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh", "unreachable": true}
2016/05/21 21:56:55 ui:     amazon-ebs: 	to retry, use: --limit @/Users/khoanguyen/packer-ebs/playbook.retry
    amazon-ebs: 	to retry, use: --limit @/Users/khoanguyen/packer-ebs/playbook.retry
```

**Comment by rickard-von-essen** [comment]
If you run with `--debug` and just after it connects to ssh
```
==> amazon-ebs: Waiting for instance (i-c8bb5e54) to become ready...
    amazon-ebs: Public DNS: ec2-52-90-8-214.compute-1.amazonaws.com
    amazon-ebs: Public IP: 52.90.8.214
    amazon-ebs: Private IP: 172.31.47.22
==> amazon-ebs: Pausing after run of step 'StepRunSourceInstance'. Press enter to continue.
==> amazon-ebs: Pausing after run of step 'stepTagEBSVolumes'. Press enter to continue.
==> amazon-ebs: Pausing after run of step 'StepGetPassword'. Press enter to continue.
==> amazon-ebs: Waiting for SSH to become available...
==> amazon-ebs: Connected to SSH!
==> amazon-ebs: Pausing after run of step 'StepConnect'. Press enter to continue.
```
It will write the private ssh key you need to connect. Try to connect manually with:
```
ssh -i ec2_amazon-ebs.pem ec2-user@52.90.8.214
```

If that doesn't work try investigate you AWS setup before shutting the packer instance. Look at VPC, ACL, SG, sub-nets etc.

**Comment by vankhoa011** [comment]
@rickard-von-essen : I can connect remote host successful by manually. Moreover, The Ansible script ran correctly by manually as well. However, the Packer still failed. 

One question, 

But I don't know why the Packer start sftp ?
==> amazon-ebs: starting sftp subsystem

I think If we use Ansible remote,  the Packer will don't need to upload Ansible playbook to remote host , Instead of, The packer will trigger Ansible from local machine

The second thing , 

What does ` unauthorized key` mean ?  I think this variable added , So these logs will not output ? 
`ansible_env_vars": ["ANSIBLE_HOST_KEY_CHECKING=False"]`

```
2016/05/21 21:56:52 ui: ==> amazon-ebs: authentication attempt from 127.0.0.1:56496 to 127.0.0.1:56495 as ec2-user using publickey
==> amazon-ebs: authentication attempt from 127.0.0.1:56496 to 127.0.0.1:56495 as ec2-user using publickey
2016/05/21 21:56:52 ui: ==> amazon-ebs: unauthorized key
==> amazon-ebs: unauthorized key
```

**Comment by Daemoen** [comment]
@vankhoa011 The sftp subsystem is required because you cannot use scp with the ansible provisioner in packer.  There are several bugs that talk about it a bit.  Make sure that your sftp_command is aimed at the correct sftp server (default path assumes ubuntu, which is not correct for centos, for instance)

---
2016/05/21 21:56:53 ui: ==> amazon-ebs: starting sftp subsystem
==> amazon-ebs: starting sftp subsystem
2016/05/21 21:56:53 packer: 2016/05/21 21:56:53 opening new ssh session
2016/05/21 21:56:54 packer: 2016/05/21 21:56:54 starting remote command: /usr/lib/sftp-server -e
**2016/05/21 21:56:55 packer: 2016/05/21 21:56:55 remote command exited with '127': /usr/lib/sftp-server -e**
2016/05/21 21:56:55 packer: 2016/05/21 21:56:55 [INFO] RPC endpoint: Communicator ended with: 127
---
This means that the sftp server failed to start.  You need to set sftp_command in your json under the provisioner:
"sftp_command": "/usr/libexec/openssh/sftp-server",

as an example (for centos/amazon linux/fedora the path above is correct)

The unauthorized_key is exactly that, ansible is being told that the key you are supplying to the remote host is not authorized to connect to said remote host.

**Comment by vankhoa011** [comment]
@Daemoen @rickard-von-essen : It worked like a charm after add the command `"sftp_command": "/usr/libexec/openssh/sftp-server",` .

Thank you very much.

**Comment by itnick** [comment]
Thanks @vankhoa011 for creating this issue.
And @Daemoen @rickard-von-essen 
Spent a morning to debug why the same playbook work on Ubuntu Linux but Amazon Linux

**Comment by rickard-von-essen** [comment]
To follow up on this, I just merged #3861 into master which adds scp support to the ansible proxy. So now you can run ansible without sftp-server.
